Pegasus/include/BoundingVolumes.hpp

```diff
@@ -133,6 +141,471 @@ class BoundingSphere
     );
 };
 } // namespace sphere
+
+namespace hierarchy
+{
+
+/**
+ * Extracts Box from AxisAlignedBoundingBox instance.
+ *
+ * @param aabb AxisAlignedBoundingBox instance to extract volume from
+ * @return extracted Box
+ */
+inline Box GetSimpleShape(aabb::AxisAlignedBoundingBox const& aabb)
+{
+    return aabb.GetBox();
+}
+
+/**
+ * Extracts Box from OrientedBoundingBox instance.
+ *
+ * @param obb OrientedBoundingBox instance to extract volume from
+ * @return extracted Box
+ */
+inline Box GetSimpleShape(obb::OrientedBoundingBox const& obb)
+{
+    return obb.GetBox();
+}
+
+/**
+ * Extracts Sphere from BoundingSphere instance.
+ *
+ * @param sphere BoundingSphere instance to extract volume from
+ * @return extracted Sphere
+ */
+inline Sphere GetSimpleShape(sphere::BoundingSphere const& boundingSphere)
+{
+    return boundingSphere.GetSphere();
+}
+
+/**
+ * General bounding volume hierarchy calculation class, builds a binary tree of bounding volumes.
+ *
+ * @tparam BoundingVolume bounding volume type to be used in hierarchy
+ */
+template <typename BoundingVolume>
+class BoundingVolumeHierarchy
+{
+public:
+    std::size_t static const MAX_NODE_SIZE = 5;
+
+    /**
+     * BoundingVolumeHierarchy internal node data structure.
+     *
+     * Cannot be copied or moved. Can only be constructed inside BoundingVolumeHierarchy class.
+     */
+    class Node
+    {
+    public:
+        BoundingVolume const volume;
+
+        Node(Node const&) = delete;
+        Node(Node &&) = delete;
+        Node & operator=(Node const&) = delete;
+        Node & operator=(Node &&) = delete;
+
+        /**
+         * Gets a lower child of this node.
+         *
+         * @return lower child
+         */
+        std::unique_ptr<Node> const& GetLowerChild() const
+        {
+            return m_lowerChild;
+        }
+
+        /**
+         * Gets an upper child of this node.
+         *
+         * @return upper child
+         */
+        std::unique_ptr<Node> const& GetUpperChild() const
+        {
+            return m_upperChild;
+        }
+
+        /**
+         * Checks if this node is a leaf, i.e. doesn't have any children.
+         *
+         * @return true if this node is a leaf, false otherwise
+         */
+        bool IsLeaf() const
+        {
+            return m_isLeaf;
+        }
+
+        ~Node() = default;
+
+    private:
+        mutable std::unique_ptr<Node> m_lowerChild;
+        mutable std::unique_ptr<Node> m_upperChild;
+        bool m_isLeaf;
+
+        /**
+         * Constructs a node out of volume. Children are initialized
+         * to nullptr, and the node is assumed to be a leaf.
+         *
+         * @param volume
```

Comment:
param with [in/out] also specify what a volume

Reply:
Same, here we have an rvalue reference, it doesn't really makes sense to mark it either `in` or `out`, we just pass ownership of this object.